File: db4oj/core/src/com/db4o/cs/messages/MWriteBatchedMessages.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.cs.messages;

import com.db4o.cs.YapServerThread;
import com.db4o.inside.*;

public class MWriteBatchedMessages extends MsgD {
	public final boolean processAtServer(YapServerThread serverThread) {
		int count = readInt();
		Transaction ta = transaction();
		for (int i = 0; i < count; i++) {
			StatefulBuffer writer = _payLoad.readYapBytes();
			int messageId = writer.readInt();
			Msg message = Msg.getMessage(messageId);
			Msg clonedMessage = message.clone(ta);
			if (clonedMessage instanceof MsgD) {
				MsgD mso = (MsgD) clonedMessage;
				mso.payLoad(writer);
				if (mso.payLoad() != null) {
					mso.payLoad().incrementOffset(Const4.MESSAGE_LENGTH - Const4.INT_LENGTH);
					mso.payLoad().setTransaction(ta);
					mso.processAtServer(serverThread);
				}
			} else { // Msg
				 if(!clonedMessage.processAtServer(serverThread)) {
					 // if the message is not processed in Msg.processAtServer
					 serverThread.processSpecialMsg(clonedMessage);
				 }
			}
		}
		return true;
	}

}

File: db4o.net/Db4objects.Db4o.CS/Db4objects.Db4o.CS/CS/Internal/Messages/MWriteBatchedMessages.cs
/* Copyright (C) 2004 - 2008  Versant Inc.  http://www.db4o.com */

using Db4objects.Db4o.CS.Internal.Messages;
using Db4objects.Db4o.Internal;

namespace Db4objects.Db4o.CS.Internal.Messages
{
	public class MWriteBatchedMessages : MsgD, IMessageWithResponse
	{
		public bool ProcessAtServer()
		{
			int count = ReadInt();
			Transaction ta = Transaction();
			lock (StreamLock())
			{
				for (int i = 0; i < count; i++)
				{
					StatefulBuffer writer = _payLoad.ReadStatefulBuffer();
					int messageId = writer.ReadInt();
					Msg message = Msg.GetMessage(messageId);
					Msg clonedMessage = message.PublicClone();
					clonedMessage.SetMessageDispatcher(MessageDispatcher());
					clonedMessage.SetTransaction(ta);
					if (clonedMessage is MsgD)
					{
						MsgD msgd = (MsgD)clonedMessage;
						msgd.PayLoad(writer);
						if (msgd.PayLoad() != null)
						{
							msgd.PayLoad().IncrementOffset(Const4.IntLength);
							Transaction t = CheckParentTransaction(ta, msgd.PayLoad());
							msgd.SetTransaction(t);
							((IServerSideMessage)msgd).ProcessAtServer();
						}
					}
					else
					{
						((IServerSideMessage)clonedMessage).ProcessAtServer();
					}
				}
				return true;
			}
		}
	}
}